src/stream/src/executor/mview/materialize.rs

```diff
@@ -558,189 +558,161 @@ pub struct MaterializeCache<SD> {
 type CacheValue = Option<CompactedRow>;
 
 impl<SD: ValueRowSerde> MaterializeCache<SD> {
-    pub fn new(
+    fn new(
         watermark_sequence: AtomicU64Ref,
         metrics_info: MetricsInfo,
         row_serde: BasicSerde,
         version_column_index: Option<u32>,
     ) -> Self {
-        let cache: ManagedLruCache<Vec<u8>, CacheValue> =
+        let lru_cache: ManagedLruCache<Vec<u8>, CacheValue> =
             ManagedLruCache::unbounded(watermark_sequence, metrics_info.clone());
         Self {
-            data: cache,
+            lru_cache,
             row_serde,
             version_column_index,
             _serde: PhantomData,
         }
     }
 
-    pub async fn handle<S: StateStore>(
+    async fn handle<S: StateStore>(
         &mut self,
         row_ops: Vec<(Op, Vec<u8>, Bytes)>,
         table: &StateTableInner<S, SD>,
-        conflict_behavior: &ConflictBehavior,
+        conflict_behavior: ConflictBehavior,
         metrics: &MaterializeMetrics,
-    ) -> StreamExecutorResult<MaterializeBuffer> {
+    ) -> StreamExecutorResult<ChangeBuffer> {
         let key_set: HashSet<Box<[u8]>> = row_ops
             .iter()
             .map(|(_, k, _)| k.as_slice().into())
             .collect();
+
+        // Populate the LRU cache with the keys in input chunk.
+        // For new keys, row values are set to None.
         self.fetch_keys(
             key_set.iter().map(|v| v.deref()),
             table,
             conflict_behavior,
             metrics,
         )
         .await?;
-        let mut fixed_changes = MaterializeBuffer::new();
+
+        let mut change_buffer = ChangeBuffer::new();
         let row_serde = self.row_serde.clone();
         let version_column_index = self.version_column_index;
-        for (op, key, value) in row_ops {
-            let mut update_cache = false;
-            let fixed_changes = &mut fixed_changes;
-            // When you change the fixed_changes buffer, you must mark `update_cache` to true so the cache and downstream can be consistent.
-            let fixed_changes = || {
-                update_cache = true;
-                fixed_changes
-            };
```

Comment:
Get rid of the clever `update_cache` variable, instead, explicitly update the cache right after `change_buffer` mutation.